### PR TITLE
Sparse cert creation for `kubeadm init`

### PR DESCRIPTION
**Pull request by liztio**
**What type of PR is this?**

> /kind bug

**What this PR does / why we need it**:

"Sparse certs" refers to the scenario where some, but not all, kubeadm certs are pre-populated. 

This PR solves the issues of #71150 in a much more robust way. The logic of whether a cert needs to created is decided within each individual phase. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #71150 kubernetes/kubeadm#918

**Special notes for your reviewer**:

A lot of the moving around in this pull request comes from me not wanting to duplicate the test cases and infra of the existing functionality (which is still used elsewhere). 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by liztio** [comment]
image

File: cmd/kubeadm/app/cmd/phases/certs_test.go [hunk 1]
```diff
@@ -20,11 +20,13 @@ import (
 	"os"
 	"testing"
 
+	"github.com/spf13/cobra"
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
 	"k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow"
 	"k8s.io/kubernetes/cmd/kubeadm/app/phases/certs"
 	"k8s.io/kubernetes/cmd/kubeadm/app/util/pkiutil"
 	testutil "k8s.io/kubernetes/cmd/kubeadm/test"
+	certstestutil "k8s.io/kubernetes/cmd/kubeadm/test/certs"
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/certs_test.go]: so i just noticed that we are doing this pattern. :[
phases and such are eventually going to be part of the vendorable backend of kubeadm.

`cmd/kubeadm/test` hold our integration tests it should not to be imported into the "backend", it should be the other way around.

i think the location for such test utils is ideally inside the backend itself, but given we are following a previous pattern this is fine i guess.

**Review by neolit123 — APPROVED** [review]
thanks a lot @liztio 
i think this fixes the issue, as much as i can tell without testing ATM.

however, i've added one comment that i think we are doing something that we should not be doing with imports.

cc @rosti

File: cmd/kubeadm/app/cmd/phases/certs_test.go (hunk 1 above)
Comment (liztio) [reply cmd/kubeadm/app/cmd/phases/certs_test.go]: That file could be anywhere, it just happened to be where it was picked. A future PR could very easily move it I think, but that's probably out of scope

**Comment by liztio** [comment]
```
[certs] Using certificateDir folder "/etc/kubernetes/pki"
[certs] Generating "etcd/ca" certificate and key
[certs] Generating "etcd/peer" certificate and key
[certs] etcd/peer serving cert is signed for DNS names [ip-172-31-32-24 localhost] and IPs [172.31.32.24 127.0.0.1 ::1]
[certs] Generating "apiserver-etcd-client" certificate and key
[certs] Generating "etcd/healthcheck-client" certificate and key
[certs] Generating "etcd/server" certificate and key
[certs] etcd/server serving cert is signed for DNS names [ip-172-31-32-24 localhost] and IPs [172.31.32.24 127.0.0.1 ::1]
[certs] Generating "ca" certificate and key
[certs] Generating "apiserver" certificate and key
[certs] apiserver serving cert is signed for DNS names [ip-172-31-32-24 kubernetes kubernetes.default kubernetes.default.svc kubernetes.default.svc.cluster.local] and IPs [10.96.0.1 172.31.32.24]
[certs] Generating "apiserver-kubelet-client" certificate and key
[certs] Using existing front-proxy-ca certificate authority
[certs] Using existing front-proxy-client certificate and key on disk[certs] Generating "sa" key and public key
```
here's what this looks like in action

File: cmd/kubeadm/app/cmd/phases/certs_test.go (hunk 1 above)
Comment (liztio) [reply cmd/kubeadm/app/cmd/phases/certs_test.go]: Filed as kubernetes/kubeadm#1257

**Review by fabriziopandini — APPROVED** [review]
@liztio Thank you very much for this quick fix!
As discussed after release we will improve this bit of code by leveraging on the cert tree structure and trying to avoid loading certificates at every phase.
/approve
/lgtm
/milestone v1.13

**Pull request merged**